src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,348 @@
+ * Time events are always sorted in the vector by the time event
+ *
+ */
+
+/**
+ * @brief xi_swap_time_events
+ *
+ * Swaps two vector xi_time_event_t elements pointed by lhs_index and rhs_index indexes
+ * with each other. It updates element's positions value to the new ones.
+ *
+ * @param vector
+ * @param lhs_index
+ * @param rhs_index
+ */
+static void xi_swap_time_events( xi_vector_t* vector,
+                                 xi_vector_index_type_t lhs_index,
+                                 xi_vector_index_type_t rhs_index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( lhs_index < vector->elem_no );
+    assert( rhs_index < vector->elem_no );
+
+    xi_time_event_t* lhs_time_event =
+        ( xi_time_event_t* )vector->array[lhs_index].selector_t.ptr_value;
+    xi_time_event_t* rhs_time_event =
+        ( xi_time_event_t* )vector->array[rhs_index].selector_t.ptr_value;
+
+    xi_vector_swap_elems( vector, lhs_index, rhs_index );
+
+    lhs_time_event->position = rhs_index;
+    rhs_time_event->position = lhs_index;
+}
+
+/**
+ * @brief xi_time_event_bubble_and_sort_down
+ *
+ * Helper function used for insertion and re-insertion ( cancel opertaion ). It uses the
+ * bubble sort idea to put the element on the proper position. In order to leave the
+ * container sorted.
+ *
+ * @param vector
+ * @param index
+ * @return new index of the element after bubbling it
+ */
+xi_vector_index_type_t
+xi_time_event_bubble_and_sort_down( xi_vector_t* vector, xi_vector_index_type_t index )
+{
+    assert( NULL != vector );
+    assert( index >= 0 );
+    assert( index < vector->elem_no );
+
+    /* prepare the tmp variables that will keep the indexes */
+    xi_vector_index_type_t elem_index        = index;
+    xi_vector_index_type_t elem_to_cmp_index = elem_index - 1;
+
+    /* let's bubble up the element to the proper position */
+    while ( elem_to_cmp_index >= 0 )
+    {
+        const xi_time_event_t* elem_to_cmp =
+            ( xi_time_event_t* )vector->array[elem_to_cmp_index].selector_t.ptr_value;
+        const xi_time_event_t* elem_to_bubble =
+            ( xi_time_event_t* )vector->array[elem_index].selector_t.ptr_value;
+
+        if ( elem_to_cmp->time_of_execution > elem_to_bubble->time_of_execution )
+        {
+            /* if the elements are not in the right order lets swap them */
+            xi_swap_time_events( vector, elem_to_cmp_index, elem_index );
+
+            /* update the indexes */
+            elem_index -= 1;
+            elem_to_cmp_index -= 1;
+        }
+        else
+        {
+            /* thanks to the container invariant on the order of the elemements we can
+             * break at this moment */
+            break;
+        }
+    }
+
+    return elem_index;
+}
+
+/**
+ * @brief xi_time_event_move_to_the_end
+ *
+ * Helper function that moves the element form the given position ( index ) to the end of
+ * the vector so it is the last element in the vector.
```

Comment:
You can remove "so it is the last element in the vector".  I think that's already covered in the terminology "to the end of the vector" in the previous clause.